### PR TITLE
Objects from Rivet-based ParticleLevelProducer

### PR DESCRIPTION
**Pull request by mseidel42**
See #8

### CODE REVIEW

**Comment by arizzi** [comment]
did you look at the size it takes?
you could  run on few k events, run treesize.py on the lzma file and post the resulting html somewhere

**Comment by intrepid42** [comment]
RivetDressedLepton (0.6 items/evt, 0.087 kb/evt)
branch	kind	b/event	b/item	plot	%
RivetDressedLepton_pt	Float_t	16.0	26.7 		17.9%
RivetDressedLepton_eta	Float_t	15.9	26.5 		17.8%
RivetDressedLepton_phi	Float_t	15.8	26.3 		17.7%
RivetDressedLepton_mass	Float_t	15.1	25.2 		16.9%
RivetDressedLepton_pdgId	Int_t	14.8	24.7 		16.6%
nRivetDressedLepton	UInt_t	11.8	19.7 		13.2%

RivetFatJet (0.4 items/evt, 0.076 kb/evt)
branch	kind	b/event	b/item	plot	%
RivetFatJet_mass	Float_t	13.6	34.0 		17.5%
RivetFatJet_eta	Float_t	13.5	33.8 		17.4%
RivetFatJet_phi	Float_t	13.5	33.8 		17.4%
RivetFatJet_pt	Float_t	13.4	33.5 		17.3%
RivetFatJet_pdgId	Int_t	12.8	32.0 		16.5%
nRivetFatJet	UInt_t	10.8	27.0 		13.9%

RivetJet (4.5 items/evt, 0.142 kb/evt)
branch	kind	b/event	b/item	plot	%
RivetJet_pt	Float_t	30.8	6.8 		21.2%
RivetJet_phi	Float_t	30.2	6.7 		20.8%
RivetJet_eta	Float_t	29.3	6.5 		20.2%
RivetJet_mass	Float_t	26.7	5.9 		18.4%
RivetJet_pdgId	Int_t	17.2	3.8 		11.8%
nRivetJet	UInt_t	11.2	2.5 		7.7%

RivetMET (1.0 items/evt, 0.024 kb/evt)
branch	kind	b/event	b/item	plot	%
RivetMET_phi	Float_t	12.1	12.1 		50.2%
RivetMET_pt	Float_t	12.0	12.0 		49.8%

RivetTag (4.7 items/evt, 0.144 kb/evt)
branch	kind	b/event	b/item	plot	%
RivetTag_pt	Float_t	31.6	6.7 		21.4%
RivetTag_phi	Float_t	30.4	6.5 		20.6%
RivetTag_eta	Float_t	30.1	6.4 		20.4%
RivetTag_pdgId	Int_t	22.3	4.7 		15.1%
RivetTag_mass	Float_t	21.7	4.6 		14.7%
nRivetTag	UInt_t	11.4	2.4 		7.7%

**Comment by arizzi** [comment]
this seem a lot of space

**Comment by arizzi** [comment]
how many events is this?

**Comment by intrepid42** [comment]
I checked 10000 events.

It would be possible to drop the RivetTag collection as those particles should already be in GenPart (although with lower precision).

**Comment by arizzi** [comment]
it seems smaller in my tests : 
http://arizzi.web.cern.ch/arizzi/nanoaod/rivet.html#RivetDressedLepton

**Comment by arizzi** [comment]
what is exactly RivetTag ?

**Comment by intrepid42** [comment]
Ok, I get the same (smaller) numbers on the lzma.root.

The RivetTag particles are the decayed "ghost" charm/bottom hadrons and taus that were clustered into the jet for the flavor definition (in case of b/c, tau is stored just for information).

They are currently used in the TOP group to derive a systematic uncertainty on b fragmentation (reweighting `x_B = pt(B hadron)/pt(jet)`). Probably the precision of the GenParts is good enough for that purpose.

**Comment by arizzi** [comment]
I'm not sure I understand how the information stored in RivetTag is supposed to be used, the x_B you mention seem single float that you can attach to the jets

**Comment by arizzi** [comment]
it seems to me that the content for example of RiveJet is a subset of the GenJets with kinematics differences that are below the precision we use to store pt,eta,phi.
Are you expecting this to be not the case in some events? or this is more or less what you expect?

**Comment by intrepid42** [comment]
> I'm not sure I understand how the information stored in RivetTag is supposed to be used, the x_B you mention seem single float that you can attach to the jets

Ok, I will remove the RivetTags, and leave it to other interested people to improve the usability...

> it seems to me that the content for example of RiveJet is a subset of the GenJets with kinematics differences that are below the precision we use to store pt,eta,phi.

That is expected. For jets, the main difference is the treatment of lepton overlap:
* RivetJets are clustered without dressed leptons (electrons or muons)
* GenJets are clustered without muons from W but including electrons.
Note that the exclusion of muons from W is not a valid observable definition at the particle level as it relies on intermediate states written in the event record.

**Comment by arizzi** [comment]
ok how about this proposal:
 - we match (by deltaR or similar) the GenJets to RivetJets
 - we do not store two collections but only one with the following additional members
    - isParticleLevel = true if GenJet is matched to RivetJets
    - deltaPt = RivetJets_pt - GenJet_pt (deltaPt is zero suppressed if < 0.something)
    - same for deltaEta,deltaPhi
    - pt is set to zero for jets that are only "rivet" (i.e. if the reverse matching fails)
- usage would be: Draw("GenJet_pt+GenJet_deltaPt","GenJet_isParticleLevel") that would do the same as  Draw("RivetJet_pt") 

This should  let us save quite some storage.

For the leptons they are small enough that they can probably stay

**Comment by intrepid42** [comment]
Seems a bit complicated to use! We would like to encourage people to use the rivet-safe collections for their measurements instead of the historic JetMET definitions. I'm a bit afraid that your proposal would result in people just using GenJets for everything, and we will have a hard time getting acceptable Rivet plugins again...

@perrozzi What do you think?

**Comment by arizzi** [comment]
we could invert the definitions but it seems RivetJets are mostly a subset of GenJets rather than viceversa...

**Comment by arizzi** [comment]
ah, that's because of the threshold... well then let's invert the definitions... @gpetruc @peruzzim  what do you think?

**Comment by perrozzi** [comment]
Hi, naively speaking I would hope that is possible to include all the rivet objects, otherwise they risk to be useless. Are you saying that lowering the threshold this is possible, and we can link genparticles or gencollections to rivet? (catching up)

**Comment by arizzi** [comment]
I'm saying we do not store two almost identical collection each taking
60bytes/events

Il 26 set 2017 7:15 PM, "perrozzi" <notifications@github.com> ha scritto:

Hi, naively speaking I would hope that is possible to include all the rivet
objects, otherwise they risk to be useless. Are you saying that lowering
the threshold this is possible, and we can link genparticles or
gencollections to rivet? (catching up)

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/cms-nanoAOD/cmssw/pull/21#issuecomment-332269637>, or mute
the thread

**Comment by perrozzi** [comment]
I agree with this. I suggest to give priority to the Rivet collection unless there are problems or features not understood. This way the gen objects analysis definitions on nanoAOD will match Rivet, and there will be full consistency with the Rivet plugins. 
Otherwise will be a mess, I am afraid. 
We can perform more detailed studies matching gen and rivet objects and trying to understand the differences, if any.

Does it sound a reasonable plan?

**Comment by arizzi** [comment]
A few points:
1)priority to rivet jets is not obvious because 
a) the current threshold is pretty high
b) we would need JME signoff on it (I understand that GEN group and some PAGs will be more happy but I'd like to hear from all)

2) from the other thread it looks like:
a) lepton selection is not agreed for all analysis so RivetJets are not a well defined concept either
b) you were (in the other thread) available to make the Jets without the lepton vetos, which would make the RivetJets and GenJets identical (the only supposed difference was actually due to a misunderstanding of the GenJet content)

3) I would not say that it is a mess to have to check one flag and make a sum for a pt correction (i.e. our original proposal) for analyses that are supposed to unfold (in addition in the postprocessing one can decide to replace Jet_pt with the Jet_pt+deltaPt)

**Comment by arizzi** [comment]
can you update the PR by commenting out the part that stores the Jets for now?
Can we schedule a discussion on this with the interested PAGs (who's that? TOP only ? SMP? )

**Comment by intrepid42** [comment]
Hi,
(including Otto, TOP Rivet contact)

I can adapt the PR on Monday, so we would store only dressed leptons and 
MET (using only particles in HF range) for now.

I would be good to have a short meeting at CERN early next week if possible!

Cheers,
Markus

**Comment by intrepid42** [comment]
Hi, this version adds the ghost tag information to the GenJet collection (as hadronFlavour) so that we can remove the RivetJets.
Unfortunately, it fails when there is a GenJet at high eta because the Rivet particleLevel can be clustered only from the particles up to |eta|<6.

Is there any use case for the jets at very high eta, or could we introduce a reasonable cut? (HF range = 5, or even a bit higher -> 5.5)

Other changes: RivetTags are removed, renamed RivetDressedLeptons->GenDressedLeptons, RivetFatJets->GenFatJets

**Comment by arizzi** [comment]
should we close and reopen on master branch?

File: PhysicsTools/NanoAOD/python/particlelevel_cff.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    cut = cms.string(""),
+    name= cms.string("GenJet"),
+    doc = cms.string("AK4 jets from Rivet-based ParticleLevelProducer"),
+    singleton = cms.bool(False), # the number of entries is variable
+    extension = cms.bool(True),
+    variables = cms.PSet(
+        # Identical to GenJets, so we just extend their flavor information
+        #P4Vars,
+        hadronFlavour = Var("pdgId", int, doc="PDG id"), 
+    )
```
Comment (arizzi) [inline PhysicsTools/NanoAOD/python/particlelevel_cff.py]: how is this aligned with the GenJet collection?

**Comment by arizzi** [comment]
even removing the jetTable that will otherwise complain because is not alligned with GenJets, I see a crash running with this file:
root://eoscms.cern.ch//eos/cms/store/mc/RunIISummer17MiniAOD/CEPDijets_M-250_13TeV-exhume/MINIAODSIM/92X_upgrade2017_realistic_v10-v1/50000/263A6592-7B93-E711-9F60-00269E95ACF8.root
<pre>
== CMSSW: Thread 1 (Thread 0x2b6ff28421a0 (LWP 257)):
== CMSSW: #0  0x00002b6ff0d67383 in poll () from /lib64/libc.so.6
== CMSSW: #1  0x00002b6ff4d2d824 in full_read.constprop () from /cvmfs/cms.cern.ch/slc6_amd64_gcc630/cms/cmssw/CMSSW_9_4_0_pre2/lib/slc6_amd64_gcc630/pluginFWCoreServicesPlugins.so
== CMSSW: #2  0x00002b6ff4d2dffa in edm::service::InitRootHandlers::stacktraceFromThread() () from /cvmfs/cms.cern.ch/slc6_amd64_gcc630/cms/cmssw/CMSSW_9_4_0_pre2/lib/slc6_amd64_gcc630/pluginFWCoreServicesPlugins.so
== CMSSW: #3  0x00002b6ff4d2f055 in sig_dostack_then_abort () from /cvmfs/cms.cern.ch/slc6_amd64_gcc630/cms/cmssw/CMSSW_9_4_0_pre2/lib/slc6_amd64_gcc630/pluginFWCoreServicesPlugins.so
== CMSSW: #4  <signal handler called>
== CMSSW: #5  0x00002b70133ece2e in GenParticles2HepMCConverter::produce(edm::Event&, edm::EventSetup const&) () from /cvmfs/cms.cern.ch/slc6_amd64_gcc630/cms/cmssw/CMSSW_9_4_0_pre2/lib/slc6_amd64_gcc630/pluginGeneratorInterfaceRivetInterface_plugins.so
== CMSSW: #6  0x00002b6fee4adb43 in edm::stream::EDProducerAdaptorBase::doEvent(edm::EventPrincipal const&, edm::EventSetup const&, edm::ActivityRegistry*, edm::ModuleCallingContext const*) () from /cvmfs/cms.cern.ch/slc6_amd64_gcc630/cms/cmssw/CMSSW_9_4_0_pre2/lib/slc6_amd64_gcc630/libFWCoreFramework.so
== CMSSW: #7  0x00002b6fee3ed462 in edm::WorkerT<edm::stream::EDProducerAdaptorBase>::implDo(edm::EventPrincipal const&, edm::EventSetup const&, edm::ModuleCallingContext const*) () from /cvmfs/cms.cern.ch/slc6_amd64_gcc630/cms/cmssw/CMSSW_9_4_0_pre2/lib/slc6_amd64_gcc630/libFWCoreFramework.so
</pre>

**Comment by arizzi** [comment]
let see if the bot would actually catch this error

**Comment by arizzi** [comment]
@intrepid42 can we use `slimmedGenJetsFlavourInfos` instead of this Rivet thing for GenJet flavour?

slimmedGenJetsFlavourInfos were introduced for this purpose in cms-sw/cmssw#19048

@Andrej-CMS can you please add in the twiki
https://twiki.cern.ch/twiki/bin/view/CMSPublic/WorkBookMiniAOD2017
an example/description on how to use the gen jets flavours?

**Comment by intrepid42** [comment]
That should work. If someone integrates it into nanoAOD I can verify that 
the definitions match for the heavy flavors.

**Comment by intrepid42** [comment]
Any news/snippets for the `slimmedGenJetsFlavourInfos`?

**Comment by Andrej-CMS** [comment]
The object is from the reco::JetFlavourInfoMatchingCollection class. So you can simply use it in the same way.

Example code is https://github.com/cms-sw/cmssw/blob/CMSSW_8_0_X/PhysicsTools/JetMCAlgos/plugins/GenHFHadronMatcher.cc

edm::Handle<reco::JetFlavourInfoMatchingCollection> jetFlavourInfos;
 evt.getByToken(jetFlavourInfosToken_, jetFlavourInfos);

look into the function std::vector<int> GenHFHadronMatcher::findHadronJets

JetFlavourInfoMatchingCollection is a map where the ->first is a reference to a jet and the ->second is a reference to the constituents that are clustered into the jet. 

const reco::GenParticleRefVector& hadronsInJet = flavour_==5 ? jetInfo.getbHadrons() : flavour_==4 ? jetInfo.getcHadrons() : reco::GenParticleRefVector();

The reference to the particles has the advantage that you can also trace the origin of the particles, if you want.

Additionally you can get the parton based flavour http://cmslxr.fnal.gov/source/PhysicsTools/JetExamples/test/printJetFlavourInfo.cc#0142

and the hadronbased flavour of the jet: http://cmslxr.fnal.gov/source/PhysicsTools/JetExamples/test/printJetFlavourInfo.cc#0106

I will update the twiki as soon as possible

**Comment by intrepid42** [comment]
Ok, I hoped that there is now something like GenJet::hadronFlavour()...

@arizzi @gpetruc Is it possible for you to add the flavor info to the MiniAOD GenJets, please?

**Comment by Andrej-CMS** [comment]
You can get the hadron based GenJetFlavour of the GenJet with the index "genJetIndex" (*jetFlavourInfos)[genJets->refAt(genJetIndex)].hadronFlavour()

This does the same thing as you are trying to do

**Comment by arizzi** [comment]
@intrepid42 nope, GenJets are not like PatJets, i.e. they cannot be extended with userfloat and at this stage of 940 release we cannot touch the dataformats. 
So you have to write a simple producer to fill that flavour information. See for example
https://github.com/cms-nanoAOD/cmssw/blob/master/PhysicsTools/NanoAOD/plugins/MuonIDTableProducer.cc
to see how one can extend an existing table

**Comment by intrepid42** [comment]
Hi all, I added jetFlavourInfos for the GenJets so that the RivetJets are not needed anymore.

However, I see that slimmedGenJetsAK8 were added, so we have the same conflict again, this time with the GenFatJet collection that I added before. Are there slimmedGenJetsAK8FlavourInfos so that we can extend the new GenJetAK8 table?

**Comment by arizzi** [comment]
mhh I'm afraid not (adding @rappoccio) ... do we really need GenJet hadron-flavour also for fatjets?

**Comment by rappoccio** [comment]
(Trying to catch up on the discussion)...

It would be great to have some kind of flavor for the subjets and/or the fat jets. It would help in b-tagging efficiency studies, QGL, etc. How big is it? Is it complicated to add?

**Comment by arizzi** [comment]
the point is that we have it for those "RivetFatJets"  but not for the Fat GenJets computed in miniaod. 
Because the two collections are very similar we need to decide a way to go. The options are the following:
a- run the gen flavour tools from mIniAOD on FatGenJets (using miniaod genparticles)
b- store those "rivet fatjets" as FatGenJets, icluding automatically the flavour
Because rivet fatjets are infact just fastjets same as our own FatGenJets it makes little difference.

**Comment by rappoccio** [comment]
Ideally synchronizing JME with the GEN RivetJets would be the way I would say we should go. However the JME conveners should weigh in here. @intrepid42 can you start an email thread to discuss this? I don't know if all of the JME conveners are using github.

**Comment by arizzi** [comment]
@zdemirag

**Comment by arizzi** [comment]
@seemasharmafnal

**Comment by intrepid42** [comment]
If @Andrej-CMS could add the corresponding JetFlavourInfoMatchingCollections in MiniAOD it would be trivial to add in nanoAOD...

However, I would be happy if we could conclude with this PR, finally. I will remove the GenFatJets (with flavor information) in favor of the GenJetAK8 collection, and we can make a new PR for adding flavor information later, favorably after having some small cofee/vidyo discussion about the way to go for that.

**Comment by arizzi** [comment]
changing the miniaod at this stage is probably already too late. If you are ok with the flavour as determined from miniaod input (I imagine for high pt jets the instabilities we suffer due to soft stuff are irrelevant) then we do not need to implement anything at miniaod level and we can just run with packedGen/prunedGen particles as input. (in addition the rivet analysis was in fact implemented with those...)

**Comment by intrepid42** [comment]
Come on, the bot complains about stuff I copied from the MuonIDTableProducer.cc!

Apart from that: Yes, the Rivet objects and their flavour information are calculated from what's available on MiniAOD, the precision should be sufficient.

**Comment by arizzi** [comment]
the bot seem to not be able to compile. does it compile for you locally
(I'm from phone right now and cannot check)

Il 31 ott 2017 4:15 PM, "Markus Seidel" <notifications@github.com> ha
scritto:

**Comment by intrepid42** [comment]
Ok, will fix it

**Comment by intrepid42** [comment]
Ok, should work now, I hope!
(Had to switch to new RelVal because of the jet (info) collections)

**Comment by arizzi** [comment]
did you also test that it works ok on80x and 92x?

I think on 80x we do not have the genjets flavour collection so we should
switch off via era customization the hadron flavor (or run on the fly the
genflavour producer)

Il 31 ott 2017 5:13 PM, "Markus Seidel" <notifications@github.com> ha
scritto:

**Comment by intrepid42** [comment]
No, it will not work on the old releases, but that's the same for the genSubJetAK8Table!

**Comment by arizzi** [comment]
for those we do have the Era switch to disable them

**Comment by intrepid42** [comment]
I see...

So it might be better to run the JetFlavourClustering on-the-fly then, which should also work in 80X and 92X. I added this (also for FatJets), and it seems to work. It just gives a lot of warnings:

```
%MSG-e MissingJetConstituent:  JetFlavourClustering:genJetFlavourAssociation  31-Oct-2017 18:48:42 CET Run: 1 E
vent: 3337079
Jet constituent required for jet reclustering is missing. Reclustered jets are not guaranteed to reproduce the 
original jets!
%MSG
%MSG-w JetPtMismatchAtLowPt:  JetFlavourClustering:genJetFlavourAssociation  31-Oct-2017 18:48:42 CET Run: 1 Ev
ent: 3337079
The reclustered and original jet 13 have different Pt's (8.68055 vs 9.31786 GeV, respectively).
Please check that the jet algorithm and jet size match those used for the original jet collection and also make
 sure the original jets are uncorrected. In addition, make sure you are not using CaloJets which are presently 
not supported.
Since the mismatch is at low Pt (Pt<10 GeV), it is ignored and only a warning is issued.

In extremely rare instances the mismatch could be caused by a difference in the machine precision in which case
 make sure the original jet collection is produced and reclustering is performed in the same job.
%MSG
```

**Comment by arizzi** [comment]
Those warnings(non fully reproducible clustering) are the reason why Andrej implemented the by value storing in miniaod.
So we need to stick to "miniaod stored by value" for 9XY releases and do the reclustering only for 80X.

If you want I can take care of the Era modifier config (but it will not happen before thursday as tomorrow is holiday here and I've another plane in 10 mins).
If you want to give it  a try yourself  you can use the "toModify" (see other examples in jets_cff.py) to change the inputTag and use the sequence cloning and module replacement/insertion (see nano_cff.py) to add the needed modules.

**Comment by intrepid42** [comment]
I think 92X also requires reclustering, or not?

What about the flavor information for the genJetsAK8? We would have to recluster them also in 94X.

**Comment by arizzi** [comment]
> I think 92X also requires reclustering, or not?
>
I'm not sure, I do not remember when Andrej PR was merged.. but if you
checked I guess you are right!


> What about the flavor information for the genJetsAK8? We would have to
> recluster them also in 94X.
>
yes, for those I imagine the instabilities are less relevant (higher pt,
hence less low pt ill defined behaviour)

**Comment by intrepid42** [comment]
Ok, using the era switch now.

(Also checked again that the 9_2_X branch does not contain Andrej's MiniAOD patch)

**Comment by Andrej-CMS** [comment]
Hello @all,

Sorry, that I'm a little slow on responding right now. I'm on vacation until Tuesday and I have rarely access to the internet.

Please let me know, if you need me to do something or if you need more information. I try to answer as fast as possible.

Kind regards,

Andrej

**Comment by arizzi** [comment]
940 version seem not to work

**Comment by intrepid42** [comment]
Fixed 2 bugs that caused crashes in 94X configuration

**Comment by intrepid42** [comment]
Uff, it seems in rare cases we have a mismatch between the original jets and the reclustering...

> Begin processing the 1640th record. Run 1, Event 3090, LumiSection 62 at 01-Nov-2017 17:24:03.636 CET
----- Begin Fatal Exception 01-Nov-2017 17:24:03 CET-----------------------
An exception of category 'LogicError' occurred while
   [0] Processing  Event run: 1 lumi: 62 event: 3087 stream: 1
   [1] Running path 'NANOAODSIMoutput_step'
   [2] Calling method for module NanoAODOutputModule/'NANOAODSIMoutput'
Exception Message:
Mismatch in number of entries between extension and main table for GenJetAK8

**Comment by arizzi** [comment]
it still fails on 80X but I do not understand why there is a mismatch if you loop on the same jet inputtag..

**Comment by intrepid42** [comment]
> it still fails on 80X but I do not understand why there is a mismatch if you loop on the same jet inputtag..

I made sure that there is a flavor entry for every GenJet passing the selection cut (if GenJet and reclustered jet don't match it falls back to `0`).

The test failure seems to be related to some DQM stuff?

> ----- Begin Fatal Exception 01-Nov-2017 20:28:43 CET-----------------------
An exception of category 'LogicError' occurred while
   [0] Processing  Event run: 1 lumi: 79 event: 3937 stream: 2
   [1] Running path 'dqmoffline_step'
   [2] Calling method for module NanoAODDQM/'nanoDQMMC'
Exception Message:
Mismatch in adding extension 
----- End Fatal Exception -------------------------------------------------

**Comment by arizzi** [comment]
it can be that the DQM fail before the OutputModule has a chance to fail, but it implements the same check (right @gpetruc?)

**Comment by gpetruc** [comment]
I think so.
can you just try run a 80X MC job with large statistics from that file?
see
https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/blob/master/scripts/run_test_long.sh#L42

**Comment by intrepid42** [comment]
Checking now with
> cmsDriver.py test_mc_94X -s NANO --mc --eventcontent NANOAODSIM --datatier NANOAODSIM --filein /store/cmst3/group/nanoAOD/CMSSW_9_4_0_pre3_RelValTTbar_13_MINIAODSIM_PU25ns_94X_mc2017_realistic_v4-v1_10000_52B94CC0-6FBB-E711-B577-0CC47A7C35F8.root --conditions auto:phase1_2017_realistic -n 10000 --era Run2_2017

But I don't understand why `test-long-mc80X-pr` succeeds, is being retried, and then fails?

**Comment by intrepid42** [comment]
Ok, so I see that 80X failed with the same error as the 94X, and then it was retried with success. So there seems to be some random effect?

My local test of 94X ran over all 5500 events of the file without problem.

**Comment by arizzi** [comment]
right, I'm also testing it and it is running fine... the fact that the behaviour is erratic is even worse... if we do not spot why it is  we need to run valgrind or stuff like that to understand the erratic behavior

**Comment by arizzi** [comment]
I rerun all pipelines to see if they fail again at the same event or it is completely random

**Comment by arizzi** [comment]
one out of  6 failed...

**Comment by arizzi** [comment]
at 4th event !

**Comment by arizzi** [comment]
I manage to reproduce it once in locale, in this crash it is due to
GenJetAK8 
size 2 vs 1

**Comment by arizzi** [comment]
@gpetruc may have found that the string parser is not thread safe, can you try switching from global producer to stream producer ?

**Comment by arizzi** [comment]
please apply the same changes as in: https://github.com/cms-nanoAOD/cmssw/pull/62/files

**Comment by arizzi** [comment]
we should clean up the commented stuff and possibly remove the word "rivet" where/if not relevant @peruzzim  or me will take care

**Comment by intrepid42** [comment]
Ok, thanks a lot!

**Pull request merged**